File: bin/boj/BOJ_11660_FindingPrefixSum_5.py
import sys


def generate_prefix_sum(matrix):
    rows, cols = len(matrix), len(matrix[0])
    prefix_sum = [[0] * cols for _ in range(rows)]

    for i in range(rows):
        for j in range(cols):
            if i > 0:
                top_sum = prefix_sum[i - 1][j]
            else:
                top_sum = 0

            if j > 0:
                left_sum = prefix_sum[i][j - 1]
            else:
                left_sum = 0

            if i > 0 and j > 0:
                diagonal_sum = prefix_sum[i - 1][j - 1]
            else:
                diagonal_sum = 0

            prefix_sum[i][j] = matrix[i][j] + top_sum + left_sum - diagonal_sum

    return prefix_sum


def get_submatrix_sum(prefix_sum, row1, col1, row2, col2):
    sum_total = prefix_sum[row2][col2]

    if row1 > 0:
        sum_total -= prefix_sum[row1 - 1][col2]

    if col1 > 0:
        sum_total -= prefix_sum[row2][col1 - 1]

    if row1 > 0 and col1 > 0:
        sum_total += prefix_sum[row1 - 1][col1 - 1]

    return sum_total


N, M = map(int, sys.stdin.readline().split())
table = [0 * N for _ in range(N)]

for i in range(N):
    table[i] = list(map(int, sys.stdin.readline().split()))

prefix_sum = generate_prefix_sum(table)

for _ in range(M):
    x1, y1, x2, y2 = map(int, sys.stdin.readline().split())
    print(get_submatrix_sum(prefix_sum, x1 - 1, y1 - 1, x2 - 1, y2 - 1))

File: src/boj/BOJ_11660_FindingPrefixSum_5.java
package boj;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class BOJ_11660_FindingPrefixSum_5 {
    private static int[][] generatePrefixSum(int[][] matrix) {
        int rows = matrix.length;
        int cols = matrix[0].length;
        int[][] prefixSum = new int[rows][cols];

        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                int topSum = (i > 0) ? prefixSum[i - 1][j] : 0;
                int leftSum = (j > 0) ? prefixSum[i][j - 1] : 0;
                int diagonalSum = (i > 0 && j > 0) ? prefixSum[i - 1][j - 1] : 0;

                prefixSum[i][j] = matrix[i][j] + topSum + leftSum - diagonalSum;
            }
        }

        return prefixSum;
    }

    private static int getSubmatrixSum(int[][] prefixSum, int row1, int col1, int row2, int col2) {
        int sumTotal = prefixSum[row2][col2];

        if (row1 > 0) {
            sumTotal -= prefixSum[row1 - 1][col2];
        }

        if (col1 > 0) {
            sumTotal -= prefixSum[row2][col1 - 1];
        }

        if (row1 > 0 && col1 > 0) {
            sumTotal += prefixSum[row1 - 1][col1 - 1];
        }

        return sumTotal;
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int N = Integer.parseInt(st.nextToken());
        int M = Integer.parseInt(st.nextToken());
        int[][] table = new int[N][N];

        for (int i = 0; i < N; i++) {
            st = new StringTokenizer(br.readLine());

            for (int j = 0; j < N; j++) {
                table[i][j] = Integer.parseInt(st.nextToken());
            }
        }

        int[][] prefixSum = generatePrefixSum(table);

        for (int k = 0; k < M; k++) {
            st = new StringTokenizer(br.readLine());
            int x1 = Integer.parseInt(st.nextToken());
            int y1 = Integer.parseInt(st.nextToken());
            int x2 = Integer.parseInt(st.nextToken());
            int y2 = Integer.parseInt(st.nextToken());
            System.out.println(getSubmatrixSum(prefixSum, x1 - 1, y1 - 1, x2 - 1, y2 - 1));
        }

        br.close();
    }
}
